File: C++ codes/EXPT 1/Transposition.cpp
#include <iostream>
#include <string>
#include <cmath>

using namespace std;

int main() {
    string a, b = "", key;
    cout << "Enter the string to be encrypted: ";
    getline(cin, a);
    cout << "Enter key: ";
    cin >> key;

    int n = key.length(); 
    int r = ceil((float)a.length() / n); 
    char m[r][n];
    for (int i = 0; i < r; i++) {
        for (int j = 0; j < n; j++) {
            if (i * n + j < a.length()) {
                m[i][j] = a[i * n + j];
            } else {
                m[i][j] = '_'; 
            };
        }
    }
    for(int i=0;i<n;i++)
    {
        int q= key[i]-'0'-1;
        for(int j=0;j<r;j++)
        {
            b+=m[j][q];
        }
    }    
    cout<<b;
}

File: Python codes/Ciphers/Transposition.py
import math

def columnar_transposition_encrypt(text, key):
    b = ""
    n = len(key)
    r = math.ceil(len(text) / n)

    # Create a matrix and fill it row-wise
    matrix = [['_' for _ in range(n)] for _ in range(r)]
    for i in range(len(text)):
        row = i // n
        col = i % n
        matrix[row][col] = text[i]

    # Build cipher text column-wise based on the key order
    for i in range(n):
        q = int(key[i]) - 1  # Convert key character to column index
        for j in range(r):
            b += matrix[j][q]

    return b

# === Main Logic ===
a = input("Enter the string to be encrypted: ")
key = input("Enter key (e.g. 3142): ")
cipher = columnar_transposition_encrypt(a, key)
print(cipher)
